apps/system/js/screen_auto_brightness.js

```diff
@@ -0,0 +1,136 @@
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_MINIMUM = 0.1;
+
+  /*
+   * This constant is used in the auto brightness algorithm. We take
+   * the base 10 logarithm of the incoming lux value from the light
+   * sensor and multiplied it by this constant. That value is used to
+   * compute a weighted average with the current brightness and
+   * finally that average brightess is and then clamped to the range
+   * [AUTO_BRIGHTNESS_MINIMUM, 1.0].
+   *
+   * Making this value larger will increase the brightness for a given
+   * ambient light level. At a value of about .25, the screen will be
+   * at full brightness in sunlight or in a well-lighted work area.
+   * At a value of about .3, the screen will typically be at maximum
+   * brightness in outdoor daylight conditions, even when overcast.
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_CONSTANT = 0.27;
+
+  /*
+   * We won't set a new brightness value if the difference between the old and
+   * new ambient light sensor values is lower than this constant.
+   */
+  ScreenAutoBrightness.prototype.AUTO_BRIGHTNESS_MIN_DELTA = 10;
+
+  /*
+   * This variable contains the latest ambient light sensor value that was used
+   * to set a brightness.
+   */
+  ScreenAutoBrightness.prototype._previousLux = undefined;
+
+  /*
+   * auto adjust screen delays, see autoAdjustBrightness function
+   * wait for _autoAdjustDelay milliseconds before adjusting brightness value
+   */
+  ScreenAutoBrightness.prototype._autoAdjustState = 0;
+  ScreenAutoBrightness.prototype._autoAdjustDelay = 750;
+
+  /*
+   * keep track of last lux value during autoAdjustBrightness delay's
+   */
+  ScreenAutoBrightness.prototype._autoDelayPrevLux = undefined;
+
+  // reinitialize to start state
+  ScreenAutoBrightness.prototype.init = function() {
+    this._previousLux = undefined;
+    this._autoAdjustState = 0;
+    this._autoDelayPrevLux = undefined;
+  };
+
+  //
+  // Automatically adjust the screen brightness based on the ambient
+  // light (in lux) measured by the device light sensor
+  // This function uses a state machine to delay before
+  // adjusting the brightness. If the user waves a finger over
+  // the sensor, the display brightness will not be changed.
+  // The state machine (recorded in _autoAdjustState):
+  // state 0: startup state, lux value is stored in _autoDelayPrevLux
+  //          transitions to state 1 if lux delta > AUTO_BRIGHTNESS_MIN_DELTA
+  // state 1: in delay timer, lux value is stored in _autoDelayPrevLux
+  //          transitions to state 2 once time delay has elapsed
+  // state 2: end of delay, use _autoDelayPrevLux value for screen brightness
+  //          transitions to state 0 automatically
+  //
+
+  ScreenAutoBrightness.prototype.autoAdjust = function(lux) {
+    var self = this;
+    var brightnessDelta;
+    if (lux < 1) { // Can't take the log of 0 or negative numbers
+      lux = 1;
+    }
+    if (this._previousLux !== undefined) {
```

Comment:
nit: Let's flip the logic here and early return, to avoid long `if` block?

```javascript
if (typeof this._previousLux === 'undefined') {
  ...
  return;
}
```

You can also create another `_setBrightnessFromLux()` function to avoid duplicate of code in this early return block and the bottom of the function.